docs/advanced-topics/docker-alternatives.rst

```diff
@@ -0,0 +1,117 @@
+`rootless Docker <https://engineering.docker.com/2019/02/experimenting-with-rootless-docker/>`_,
+and `udocker <https://indigo-dc.gitbook.io/udocker/>`_.
+While Dockstore uses Docker by default, if necessary it may be possible to run your workflows with one
+of these alternatives. Because the call to Docker or an alternative is made by the workflow runner, usually cwltool
+or Cromwell, and not Dockstore directly, the difficulty of configuring a Docker alternative depends on the workflow
+type. Some Dockstore entries will run seamlessly without Docker, and some may be entirely incompatible in a rootless
+environment.
+
+Singularity
+-----------
+
+Singularity is perhaps the most well-supported Docker alternative. Singularity can pull Docker images and build them
+into its own image format (.sif), but not all Docker features are compatible. For instance, Dockerfile ``USER``
+commands are not compatible with Singularity.
+A common problem observed when running Dockstore entries with Singularity is that the process fails on
+``singularity pull`` because the entry's Dockerfile or its base image contains a ``USER root`` command. In many cases
+the use of root may be unnecessary. Whenever possible, Dockerfiles on Dockstore should avoid using root.
+
+More information about compatibility of Dockerfiles with Singularity
+can be found `here <https://sylabs.io/guides/3.4/user-guide/singularity_and_docker.html#best-practices>`_.
+
+Singularity can be installed following the instructions
+`here <https://sylabs.io/guides/3.4/user-guide/quick_start.html>`_. Note that the installation is relatively complicated
+and requires ``sudo``.
+
+
+cwltool
+~~~~~~~
+
+Singularity is available as a command line option for cwltool.
+To set this option through Dockstore, add the following line to your ``~/.dockstore/config``:
+
+``cwltool-extra-parameters: --singularity``
+
+Cromwell
+~~~~~~~~
+
+Dockstore currently does not support custom configuration of Cromwell, so Singularity cannot be used for WDL entries
+at this point. This is a feature that we hope to add soon.
+
+Cromwell can be configured to use Singularity instead of Docker as described
+`here <https://cromwell.readthedocs.io/en/stable/tutorials/Containers/#singularity>`_.
+This requires creating a Cromwell config file with a section describing the backend provider settings.
+Examples of this are available in the Cromwell GitHub
+`here <https://github.com/broadinstitute/cromwell/tree/develop/cromwell.example.backends>`_.
+
+Rootless Docker
+---------------
+
+Rootless Docker, a product of Docker, is very convenient because no configuration of Dockstore is required to use it.
+When it is installed, all ``docker`` commands are run in rootless mode without needing to set this as an option.
+Therefore, the normal Docker commands invoked by cwltool and Cromwell will be executed rootlessly.
+
+Rootless Docker installation is simple and does not require root. Regular Docker must not be accessible in order
```

Comment:
I don't understand "Regular Docker must not be accessible...". Does that mean Docker should not be installed at all?

Reply:
I think so... even with the daemon not running, if the docker command is recognized it seems you can't install rootless docker over it.

Reply:
is this just a `PATH` thing? Like you can only have one of the two commands on your path?
Alternatively, you really can't have both rootless and normal Docker installed at the same time. 
Should clarify

Reply:
I would lean towards saying "must not be installed", or that you must install either Docker or rootless Docker (even though it gets into a grey area with Denis' examples). "Accessbile" doesn't feel like the right word to me in this context.

Reply:
The error message is "Aborting because rootful Docker is running and accessible. Set FORCE_ROOTLESS_INSTALL=1 to ignore." It says that even when docker is not on the path, so I'll just say it should not be installed.